phd-tests/tests/src/boot_order.rs

```diff
@@ -0,0 +1,673 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use anyhow::{bail, Error};
+use byteorder::{LittleEndian, ReadBytesExt};
+use phd_framework::{
+    disk::{fat::FatFilesystem, DiskSource},
+    test_vm::{DiskBackend, DiskInterface},
+};
+use phd_testcase::*;
+use std::io::{Cursor, Read};
+use tracing::{info, warn};
+
+// This test checks that with a specified boot order, the guest boots whichever disk we wanted to
+// come first. This is simple enough, until you want to know "what you booted from"..
+//
+// For live CDs, such as Alpine's, the system boots into a tmpfs loaded from a boot disk, but
+// there's no clear line to what disk that live image *came from*. If you had two Alpine 3.20.3
+// images attached to one VM, you'd ceretainly boot into Alpine 3.20.3, but I don't see a way to
+// tell *which disk* that Alpine would be sourced from, from Alpine alone.
+//
+// So instead, check EFI variables. To do this, then, we have to.. parse EFI variables. That is
+// what this test does below, but it's probably not fully robust to what we might do with PCI
+// devices in the future.
+//
+// A more "future-proof" setup might be to just boot an OS, see that we ended up in the OS we
+// expected, and check some attribute about it like that the kernel version is what we expected the
+// booted OS to be. That's still a good fallback if we discover that parsing EFI variables is
+// difficult to stick with for any reason. It has a downside though: we'd have to keep a specific
+// image around with a specific kernel version as either the "we expect to boot into this" image
+// or the "we expected to boot into not this" cases.
+
+// That all said, while this test checks for specific EDK2/OVMF magic values, those magic values
+// should be explained. The rest of this comment does just that.
+//
+// First, soem GUIDs. These GUIDs come from EDK2, and OVMF reuses them. Notably these are the raw
+// bytes of the GUID: textual values will have slightly different ordering of bytes.
+//
+// Some source references, as you won't find these GUIDs in a UEFI or related spec document.. The
+// firmware volume is identified by what seems to be the DXE firmware volume:
+// https://github.com/tianocore/edk2/blob/712797c/OvmfPkg/OvmfPkgIa32.fdf#L181
+// introduced in
+// https://github.com/tianocore/edk2/commit/16f26de663967b5a64140b6abba2c145ea50194c, note this
+// is the DXEFV entry.
+//
+// The *files* we'll care about in this test are identified by other GUIDs in the above *volume*.
+//
+// EFI Internal Shell: https://github.com/tianocore/edk2/blob/a445e1a/ShellPkg/ShellPkg.dec#L59-L60
+// UiApp:
+// https://github.com/tianocore/edk2/blob/a445e1a/MdeModulePkg/Application/UiApp/UiApp.inf#L13
```

Comment:
i was initially concerned went overboard here, but this turned out a very useful detail to be checking. i feel much better about this test version of the test now than i did when starting to write it last week :sweat_smile: